apps/authentication-service/src/shared/http/httpService.ts

```diff
@@ -0,0 +1,142 @@
+import { ResponseError } from "auth0";
+import fetchIntercept, { FetchInterceptorResponse } from "fetch-intercept";
+
+import { AuthService, authEndpoint } from "@/shared/auth/authService";
+import { IAuthService } from "@/shared/auth/authModel";
+import {
+  IHttpDataRequest,
+  IHttpRefreshCallback,
+  IHttpRequestOptions,
+  IHttpService,
+} from "@/shared/http/httpModel";
+import { ISingleton } from "@/shared/models";
+
+const getRelativeUrl = (url: string): string => {
+  const _url = new URL(url);
+
+  return url.replace(_url.origin, "");
+};
+
+export const parseResponse = async (response: Response) => {
+  if (!response.ok) {
+    throw Error(response.statusText);
+  }
+  return response.json();
+};
+
+export const HttpService: ISingleton<IHttpService> = (() => {
+  let instance: IHttpService;
+  let Auth: IAuthService;
+
+  let refreshSubscribers: IHttpRefreshCallback[] = [];
+  let dataRequests = {} as IHttpDataRequest;
+
+  const subscribeTokenRefresh = (callback: IHttpRefreshCallback): void => {
+    refreshSubscribers.push(callback);
+  };
+
+  const onRefreshed = () => {
+    refreshSubscribers.map((callback: IHttpRefreshCallback) => callback());
+    refreshSubscribers = [];
+  };
+
+  const removeDataRequestsItem = (requestKey: string) => {
+    const { [requestKey]: _omit, ...remaining } = dataRequests;
+    dataRequests = remaining;
+  };
+
+  const createInstance = (): IHttpService => {
+    let isRefreshing = false;
+    Auth = AuthService.getInstance();
+
+    const unregisterInterceptor = fetchIntercept.register({
+      request: (url, config: Request): IHttpRequestOptions => {
+        console.log(`Intercepted request: ${url}, method: ${config.method}`);
+
+        if (config && url.includes(authEndpoint)) {
+          dataRequests = {
+            ...dataRequests,
+            [`${getRelativeUrl(url)}_${config.method || "GET"}`]: config,
+          };
+        }
+
+        return [url, config];
+      },
+
+      response: (
+        response: FetchInterceptorResponse,
+      ): FetchInterceptorResponse => {
+        console.log(
+          `Intercepted response: ${response.url}, method: ${response.request.method}, status: ${response.status}`,
+        );
+
+        const requestKey: string = `${getRelativeUrl(response.url)}_${response.request.method}`;
+
+        if (response.status === 401 && response.url.includes(authEndpoint)) {
+          if (!isRefreshing) {
+            isRefreshing = true;
+            Auth.fetchAccessToken()
+              .then(() => {
+                onRefreshed();
+              })
+              .catch(() => {
+                // TODO: Handle logging
```

Comment:
If we can't do this now, then I think it's better to create a ticket than add a TODO comment

Reply:
Service removed